src/vsr/journal.zig

```diff
@@ -1356,93 +1355,141 @@ pub fn JournalType(comptime Replica: type, comptime Storage: type) type {
+            var torn_slot_count: u8 = 0;
+
+            // We now search for torn prepares between op_max and op_prepare_max. A torn prepare
+            // manifests as a prepare with an *invalid checksum* and a *valid* header from any
+            // previous wrap. If our op_max is certain, i.e. we are guaranteed to not find any
+            // op > op_max in our journal, then we can say with certainty that a torn prepare was
+            // being appended to the WAL. However, if we find a "non torn-prepare" fault outside of
+            // [op_max + 1, op_prepare_max], we return an empty slice.
             //
-            //   (fault  [checkpoint..........torn]        fault)
-            //   (...torn]    fault     fault  [checkpoint......)
+            //   (fault  [op_max+1..........op_prepare_max]        fault)
+            //   (...op_prepare_max]    fault     fault  [op_max+1......)
             //
-            // When there is a fault between the checkpoint and the torn prepare, we cannot be
-            // certain if the prepare was truly torn (safe to truncate) or corrupted (not safe to
-            // truncate).
-            //
-            // When the checkpoint and torn op are in the same slot, then we can only be certain
-            // if there are no faults other than the torn op itself.
+            // When there exists a "non torn-prepare" fault outside of [op_max + 1, op_prepare_max],
+            // op_max is not certain, as the faulty slot could be the true op_max. Consequently, we
+            // can't say if a torn prepare was truly torn (safe to truncate) or corrupted (not safe
+            // to truncate).
             for (cases, 0..) |case, index| {
                 // Do not use `faulty.bit()` because the decisions have not been processed yet.
                 if (case.decision(replica.solo()) == .vsr) {
-                    if (checkpoint_index == torn_slot.index) {
-                        assert(op_max >= replica.op_checkpoint());
-                        assert(torn_op > replica.op_checkpoint());
-                        if (index != torn_slot.index) return null;
-                    } else {
-                        if (!known_range.contains(Slot{ .index = index })) return null;
+                    const slot = Slot{ .index = index };
+
+                    // We only need to check the slot for op_prepare_max. Checked separately as
+                    // SlotRange.contains doesn't handle empty ranges.
+                    if (op_max == op_prepare_max - 1 and index != op_prepare_max_slot.index) {
+                        assert(op_max_to_op_prepare_max.head.index == op_prepare_max_slot.index);
+                        assert(op_max_to_op_prepare_max.tail.index == op_prepare_max_slot.index);
+                        continue;
+                    }
+
+                    if ((op_max == op_prepare_max - 1 and index == op_prepare_max_slot.index) or
+                        op_max_to_op_prepare_max.contains(slot))
```

Comment:
Extract `op_max == op_prepare_max - 1` to a `range_empty` variable, for clarity. (Also, should it use `op_max_to_op_prepare_max.tail` instead of `op_prepare_max - 1`?)

Reply:
Took care of this in https://github.com/tigerbeetle/tigerbeetle/pull/2345/commits/8c0600bcefe1bab79a1df21d142fbe2c3d4a10d3.